features/64-chat-offline-cache/api.md

```diff
@@ -0,0 +1,373 @@
+func fetchMessages(conversation, beforeMessage, limit, order, cachedCallback?)
+func fetchMessages(conversation, afterMessage, limit, order, cachedCallback?)
+```
+
+#### JS
+
+```js
+const successCallback = (conversations, cached = false) => {
+  // display conversations
+};
+
+skygearChat
+  .fetchConversations(successCallback)
+  .then(successCallback, errorCallback);
+
+// no-cache version
+skygearChat
+  .fetchConversation()
+  .then(successCallback, errorCallback);
+```
+
+```js
+let currentMessages = [];
+
+const successCallback = (messages, cached = false) => {
+  currentMessages = skygearChat.merge(currentMessages, messages);
+  // display currentMessages
+};
+
+skygearChat
+  .fetchMessages(conversation, beforeMessage, limit, order, successCallback)
+  .then(successCallback, errorCallback);
+
+// no-cache version
+skygearChat
+  .fetchMessages(conversation, beforeMessage, limit, order)
+  .then(successCallback, errorCallback);
+```
+
+#### Non-JS
+
+```
+skygearChat.fetchConversations(func (conversations, cached = false) {
+  // display conversations
+})
+```
+
+```
+let currentMessages = [];
+
+skygearChat.fetchMessages(func (messages, cached = false) {
+  currentMessages = skygearChat.merge(currentMessages, messages);
+  // display currentMessages
+})
+```
+
+- Add utility functions for handling result returned by chat container
+
+```
+func merge(conversationDeltas: [ConversationDelta], toList: [Conversation]) -> [Conversation]
+func merge(messages: [Message], toList: [Message]) -> [Message]
+func merge(messageDeltas: [MessageDelta], toList: [Message]) -> [Message]
+
+func findBoundaries(ofMessageList: [Message]) -> [(Int, Int)]
+```
+
+- Add a managed message list for user to use directly
+
+The ManagedMessageList is a default implementation provided by Skygear, its underlying implementation is just using the above new API and utility functions.
+
+```
+class ManagedMessageList {
+
+  constructor(conversation: Conversation,
+              delegate: ManagedMessageListDelegate)
+
+  - fetchLimit: Int = 100
+  - useCache: Boolean = true
+  - subscribeNewMessages: Boolean = true
+  - syncOnReconnect: Boolean = true
+
+  func online() // expected to call when enter view and pubsub connected
+
+  func offline() // expected to call when leave view and pubsub disconnected
+
+  func messageCount() -> Int
+
+  func messageAt(position: Int) -> Message
+
+  /**
+   * scrollViewDelegateMethods
+   *
+   * the list will determine when to fetch messages by checking current first
+   * and last visible cell
+   */
+
+}
+
+interface ManagedMessageListDelegate {
+
+  func managedMessageListDidChange(messageDeltas: [MessageDelta])
+
+}
+```
+
+# Example
+
+*All below UI update, I use `this.tableView.reloadTable()`. It is up to client app to decide how to update the UI*
+
+### In conversation list
+
+#### Start subscribing conversation changes
+
+When
+- enter conversation list
+- come offline and online again when in conversation list
```

Comment:
For Pubsub, my understanding is if you called the subscribe functions, it will handle auto-reconnect for you.

So for `subscribeConversations` and `subscribeMessages`, are they also going to auto-reconnect?

If not, it might be hard to make sure user understand it?

If yes, what is "come offline and online again when in conversation list"? Maybe you meant "init the conversation list"?

Reply:
I think it is a bit controversial... maybe need a discussion to see if subscribeConversation / subscribeMessages should behave like subscribe in pubsub.

Reply:
> So for subscribeConversations and subscribeMessages, are they also going to auto-reconnect?

I think it makes sense. So the two subscription functions should be use regardless of pubsub connection state.